repository/orm.go

```diff
@@ -0,0 +1,123 @@
+package repository
+
+import (
+	"database/sql/driver"
+	"encoding/json"
+	"fmt"
+
+	"github.com/soundcloud/periskop/metrics"
+	"gorm.io/gorm"
+)
+
+type ormRepository struct {
+	DB *gorm.DB
+	targetsRepository
+}
+
+func (e *ErrorAggregate) Scan(src interface{}) error {
+	return json.Unmarshal([]byte(src.(string)), &e)
+}
+
+func (e ErrorAggregate) Value() (driver.Value, error) {
+	val, err := json.Marshal(e)
+	return string(val), err
+}
+
+type AggregatedError struct {
+	gorm.Model
+	ServiceName    string
+	AggregationKey string
+	Errors         ErrorAggregate
+}
+
+func NewORMRepository(db *gorm.DB) ErrorsRepository {
+	err := db.AutoMigrate(&AggregatedError{})
+	if err != nil {
+		panic("failed to create database migration")
+	}
+	return &ormRepository{DB: db}
+}
+
+func (r *ormRepository) GetErrors(serviceName string, numberOfErrors int) ([]ErrorAggregate, error) {
+	aggregatedErrors := []AggregatedError{}
+	r.DB.
+		Where(&AggregatedError{ServiceName: serviceName}).
+		Find(&aggregatedErrors)
+
+	errors := []ErrorAggregate{}
+	for _, aggregatedError := range aggregatedErrors {
+		errorObj := aggregatedError.Errors
+		maxErrors := len(errorObj.LatestErrors)
+		if numberOfErrors < maxErrors {
+			maxErrors = numberOfErrors
+		}
+		errorObj.LatestErrors = errorObj.LatestErrors[0:maxErrors]
+		errors = append(errors, errorObj)
+	}
+	if len(errors) > 0 {
+		return errors, nil
+	}
+	metrics.ServiceErrors.WithLabelValues("service_not_found").Inc()
+	return nil, fmt.Errorf("service %s not found", serviceName)
+}
+
+// StoreErrors deletes previous stored errors for a service name and store the new list of errors in json format
+func (r *ormRepository) StoreErrors(serviceName string, errors []ErrorAggregate) {
```

Comment:
```suggestion
func (r *ormRepository) ReplaceErrors(serviceName string, errors []ErrorAggregate) {
```
given how this function works, would it make sense to rename to make functionality clearer?

Reply:
Maybe `StoreAndReplaceErrors`?